src/track/trackmetadatataglib.cpp

```diff
@@ -564,44 +590,104 @@ void writeTrackMetadataIntoTag(
     }
 
-    // Set the numeric year if available
-    const QDate yearDate(
-            TrackMetadata::parseDateTime(trackMetadata.getYear()).date());
-    if (yearDate.isValid()) {
-        pTag->setYear(yearDate.year());
+    // Specialized write functions for tags derived from Taglib::Tag might
+    // be able to write the complete string from trackMetadata.getYear()
+    // into the corresponding field. In this case parsing the year string
+    // here should be omitted.
+    if (0 == (writeMask & WRITE_TAG_OMIT_YEAR)) {
+        // Set the numeric year if available
+        const QDate yearDate(
+                TrackMetadata::parseDateTime(trackMetadata.getYear()).date());
+        if (yearDate.isValid()) {
+            pTag->setYear(yearDate.year());
+        }
+    }
+
+    // The numeric track number in TagLib::Tag does not reflect the total
+    // number of tracks! Specialized write functions for tags derived from
+    // Taglib::Tag might be able to handle both trackMetadata.getTrackNumber()
+    // and trackMetadata.getTrackTotal(). In this case parsing the track
+    // number string here is useless and should be omitted.
+    if (0 == (writeMask & WRITE_TAG_OMIT_TRACK_NUMBER)) {
+        // Set the numeric track number if available
+        TrackNumbers parsedTrackNumbers;
+        const TrackNumbers::ParseResult parseResult =
+                TrackNumbers::parseFromString(trackMetadata.getTrackNumber(), &parsedTrackNumbers);
+        if (TrackNumbers::ParseResult::VALID == parseResult) {
+            pTag->setTrack(parsedTrackNumbers.getActual());
+        }
+    }
+}
+
+void writeMP4Atom(
+        TagLib::MP4::Tag* pTag,
+        const TagLib::String& key,
+        const QString& str) {
+    if (str.isEmpty()) {
+        // Purge empty atoms
+        pTag->itemListMap().erase(key);
+    } else {
+        TagLib::StringList strList(toTagLibString(str));
+        pTag->itemListMap()[key] = std::move(strList);
     }
-    // Derived tags might be able to write the complete string
-    // from trackMetadata.getYear() into the corresponding field.
+}
 
-    bool trackNumberValid = false;
-    uint track = trackMetadata.getTrackNumber().toUInt(&trackNumberValid);
-    if (trackNumberValid && (track > 0)) {
-        pTag->setTrack(track);
+void writeAPEItem(
+        TagLib::APE::Tag* pTag,
+        const TagLib::String& key,
+        const TagLib::String& str) {
+    if (str.isEmpty()) {
+        // Purge empty items
+        pTag->removeItem(key);
+    } else {
+        const bool replace = true;
+        pTag->addValue(key, str, replace);
     }
 }
 
-template<typename T>
-inline void writeMP4Atom(TagLib::MP4::Tag* pTag, const TagLib::String& key,
-        const T& value) {
-    pTag->itemListMap()[key] = value;
+bool readXiphCommentField(
+        const TagLib::Ogg::XiphComment& tag,
+        const TagLib::String& key,
+        QString* pValue = nullptr) {
+    if (tag.fieldListMap().contains(key)) {
```

Comment:
I mixed up reading/writing when I quickly wrote the answer. The double 
lookup is in the read function can be eliminated as you suggested :)

I will eliminate most of the double lookups by encapsulating reading/writing 
of key/value pairs from tags in separate functions. At least for the common 
cases where the type of value is a string. Double lookups for some special 
cases will remain, because it would complicate the code too much.

On 01/05/2016 10:29 PM, Daniel Schürmann wrote:
>
> In src/track/trackmetadatataglib.cpp 
> <https://github.com/mixxxdj/mixxx/pull/835#discussion_r48898646>:
>
> >      }
> >  }
> >
> > -template<typename T>
> > -inline void writeMP4Atom(TagLib::MP4::Tag* pTag, const TagLib::String& key,
> > -        const T& value) {
> > -    pTag->itemListMap()[key] = value;
> > +bool readXiphCommentField(
> > +        const TagLib::Ogg::XiphComment& tag,
> > +        const TagLib::String& key,
> > +        QString* pValue = nullptr) {
> > +    if (tag.fieldListMap().contains(key)) {
>
> I do not understand 1) please give an example.
> Are you afraid the "it" is getting invalid in my code snipped? How can 
> this happen?
> fieldListMap() just returns d->fieldListMap. If we cannot use 
> d->fieldListMap twice, some internal taglib code will fail as well.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/835/files#r48898646>.
>